File: scala/revenj-core/src/main/scala/net/revenj/database/postgres/converters/TreePathConverter.scala
package net.revenj.database.postgres.converters

import java.sql.PreparedStatement

import net.revenj.TreePath
import net.revenj.database.postgres.{PostgresBuffer, PostgresReader}
import org.postgresql.util.PGobject

object TreePathConverter extends Converter[TreePath] {
  override def serializeURI(sw: PostgresBuffer, value: TreePath): Unit = {
    if (value != null) sw.addToBuffer(value.toString)
  }

  def setParameter(sw: PostgresBuffer, ps: PreparedStatement, index: Int, value: TreePath): Unit = {
    val pg = new PGobject
    pg.setType("ltree")
    pg.setValue(value.toString)
    ps.setObject(index, pg)
  }

  def setParameter(sw: PostgresBuffer, ps: PreparedStatement, index: Int, value: Option[TreePath]): Unit = {
    val pg = new PGobject
    pg.setType("ltree")
    ps.setObject(index, pg)
  }

  val dbName = "ltree"

  def default() = TreePath.Empty

  override def parseRaw(reader: PostgresReader, start: Int, context: Int): TreePath = {
    reader.initBuffer(start.toChar)
    reader.fillUntil(',', ')')
    reader.read()
    TreePath.create(reader.bufferToString())
  }

  override def parseCollectionItem(reader: PostgresReader, context: Int): TreePath = {
    val cur = reader.read()
    reader.initBuffer(cur.toChar)
    reader.fillUntil(',', '}')
    reader.read()
    if (reader.bufferMatches("NULL")) {
      TreePath.Empty
    } else {
      TreePath.create(reader.bufferToString())
    }
  }

  override def parseNullableCollectionItem(reader: PostgresReader, context: Int): Option[TreePath] = {
    val cur = reader.read()
    reader.initBuffer(cur.toChar)
    reader.fillUntil(',', '}')
    reader.read()
    if (reader.bufferMatches("NULL")) {
      None
    } else {
      Some(TreePath.create(reader.bufferToString()))
    }
  }

  override def toTuple(value: TreePath): PostgresTuple = ValueTuple.from(value.toString)
}

File: java/revenj-core/src/main/java/org/revenj/database/postgres/converters/TreePathConverter.java
package org.revenj.database.postgres.converters;

import org.postgresql.util.PGobject;
import org.revenj.TreePath;
import org.revenj.database.postgres.PostgresBuffer;
import org.revenj.database.postgres.PostgresReader;

import java.io.IOException;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public abstract class TreePathConverter {

	public static void setParameter(PostgresBuffer sw, PreparedStatement ps, int index, TreePath value) throws SQLException {
		PGobject pg = new PGobject();
		pg.setType("ltree");
		pg.setValue(value != null ? value.toString() : null);
		ps.setObject(index, pg);
	}

	public static void serializeURI(PostgresBuffer sw, TreePath value) throws IOException {
		if (value == null) return;
		sw.addToBuffer(value.toString());
	}

	public static TreePath parse(PostgresReader reader, boolean allowNulls) throws IOException {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return allowNulls ? null : TreePath.EMPTY;
		}
		reader.initBuffer((char) cur);
		reader.fillUntil(',', ')');
		reader.read();
		return TreePath.create(reader.bufferToString());
	}

	public static List<TreePath> parseCollection(PostgresReader reader, int context, boolean allowNull) throws IOException {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return null;
		}
		boolean escaped = cur != '{';
		if (escaped) {
			reader.read(context);
		}
		cur = reader.peek();
		if (cur == '}') {
			if (escaped) {
				reader.read(context + 2);
			} else {
				reader.read(2);
			}
			return new ArrayList<>(0);
		}
		List<TreePath> list = new ArrayList<>();
		TreePath emptyCol = allowNull ? null : TreePath.EMPTY;
		do {
			cur = reader.read();
			reader.initBuffer((char) cur);
			reader.fillUntil(',', '}');
			cur = reader.read();
			if (reader.bufferMatches("NULL")) {
				list.add(emptyCol);
			} else {
				list.add(TreePath.create(reader.bufferToString()));
			}
		} while (cur == ',');
		if (escaped) {
			reader.read(context + 1);
		} else {
			reader.read();
		}
		return list;
	}

	public static PostgresTuple toTuple(TreePath value) {
		return value != null ? ValueTuple.from(value.toString()) : null;
	}
}
